src/Database/Sql/Util/Tables.hs

```diff
@@ -39,18 +39,33 @@ import qualified Data.Text.Lazy as TL
 
 import           Control.Monad.Identity
 import           Control.Monad.Writer
+import           Control.Monad.Reader
 import           Control.Monad.State
 
 import           Database.Sql.Type
 import           Database.Sql.Position
 
 
+-- This describes the usage of a particular table, whether it was read from or
+-- written to. Note that if a table is used in both modes, it will show up
+-- twice as each mode.
+data UsageMode = ReadData | ReadMeta | WriteData | WriteMeta | Unknown
```

Comment:
Oh, very interesting!

Might consider `Map FullyQualifiedTableName (Set UsageMode)` rather than the set of the pairs, but no strong feelings.